File: LeetCode/886.php
<?php
class Solution
{
    public array $map;
    public array $colors;

    /**
     * 着色法 + DFS
     * 
     * @param Integer $N
     * @param Integer[][] $dislikes
     * @return Boolean
     */
    function possibleBipartition($N, $dislikes): bool
    {
        if ($dislikes == null || count($dislikes) == 0) return true;
        $this->map = array_fill(1, $N, []);
        $this->colors = array_fill(1, $N, 0);
        foreach ($dislikes as $dislike) {
            $this->map[$dislike[0]][] = $dislike[1];
            $this->map[$dislike[1]][] = $dislike[0];
        }

        for ($i = 1; $i <= $N; $i++) {
            if ($this->colors[$i] == 0 && $this->dfs($i, 1) == false)
                return false;
        }

        return true;
    }

    /**
     * @param int $i 节点号
     * @param int $color 颜色 => 1:red | -1:blue
     */
    public function dfs(int $i, int $color): bool
    {
        $this->colors[$i] = $color;
        foreach ($this->map[$i] as $dislike) {
            if ($this->colors[$dislike] == $color) return false;
            if ($this->colors[$dislike] == 0 && $this->dfs($dislike, -$color) == false) return false;
        }
        return true;
    }
}

File: LeetCode/886.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Solution
{
    public HashMap<Integer, List<Integer>> map;
    public int[] colors;

    public boolean possibleBipartition(int N, int[][] dislikes) {
        if (dislikes == null || dislikes.length == 0) return true;
        map = new HashMap<>();
        colors = new int[N + 1];
        for (int i = 1; i <= N; i++) map.put(i, new ArrayList<>());
        for (int[] dislike : dislikes) {
            map.get(dislike[0]).add(dislike[1]);
            map.get(dislike[1]).add(dislike[0]);
        }
        for (int i = 1; i <= N; i++) if (colors[i] == 0 && !dfs(i, 1)) return false;

        return true;
    }

    private boolean dfs(int i, int color) {
        colors[i] = color;
        for (Integer index : map.get(i)) {
            if (colors[index] == color) return false;
            if (colors[index] == 0 && !dfs(index, -color)) return false;
        }
        return true;
    }

    public int[] father;

    /**
     * 并查集
     */
    public boolean possibleBipartition1(int N, int[][] dl) {
        father = new int[N * 2 + 1]; // self + dislike
        for (int i = 1; i <= N * 2; i++) father[i] = i;

        for (int i = 0; i < dl.length; i++) {
            int x = find(dl[i][0]); // 查找自己的根节点
            int y = find(dl[i][1]);
            int a = find(dl[i][0] + N); // 查找自己不喜欢的人的根节点
            int b = find(dl[i][1] + N);
            if (x == y)
                return false;
            else {
                father[a] = y; // 将不喜欢的人合并
                father[b] = x;
            }
        }
        return true;
    }

    private int find(int x) {
        if (x != father[x])
            father[x] = find(father[x]);
        return father[x];
    }

    private int find1(int x) {
        return x == father[x] ? x : (father[x] = find(father[x]));
    }
}
